feat: 사용자 프로필 이미지 호스팅 기능 추가

Wonyocho opened the pull request:
#41

Commit:
feat: 사용자 프로필 이미지 호스팅 기능 추가

Ares-10 commented on src/global/config/index.ts:
```diff
@@ -1,3 +1,6 @@
+// import dotenv from 'dotenv';
+// dotenv.config();
+
```
불필요한 주석인 것 같습니다

Wonyocho replied:
해당사항 수정했습니다~!

Commit:
feat: 불필요한 로그 제거 및 코드 정리

Ares-10 commented on src/routes/user/user.ts:
```diff
@@ -17,13 +17,13 @@ const userRoute: FastifyPluginAsync = async (fastify) => {
       const userId = request.user.id;
-      const userData = request.body as { name?: string; image?: string | null };
+      const userData = request.body as { name?: string };
       const user = await fastify.userService.updateUser(userId, userData);
```
입력값 수정에 따른 서비스 코드 내부 로직 수정이 필요할 것 같습니다.
더하여 http 메서드가 patch이며 name, image 필드 입력 받았는데 name만 입력받게 되니 Put 메서드로 바꾸고 name 필드 notnull로 하는 게 좋을 것 같습니다.

Wonyocho replied:
해당사항 수정 완료했습니다~@

Ares-10 commented on src/schemas/user/userSchema.ts:
```diff
@@ -41,6 +41,39 @@ export const updateUserSchema = {
   },
 };
 
+// 사용자 프로필 이미지 업로드 스키마
```
이미지 또한 스웨거로 입력받을 수 있는 걸로 아는데 이 부분 관련해서 찾아보시면 좋을 듯 합니다

Commit:
refac: 사용자 이름 수정 기능으로 업데이트 및 스키마 변경

Commit:
feat: multipart/form-data 처리 개선 및 프로필 이미지 업로드 스키마 수정

Wonyocho committed:
Merge branch 'main' into feat/41-profile-image-hosting

Ares-10 commented on src/main.ts:
```diff
@@ -53,16 +66,18 @@ await fastify.register(friendService);
 await fastify.register(authRoute, { prefix: '/ft/api/auth' });
 await fastify.register(twoFactorAuthRoute, { prefix: '/ft/api/auth' });
 await fastify.register(userRoute, { prefix: '/ft/api/users' });
+await fastify.register(userImageRoute, { prefix: '/ft/api/users' });
```
유저 이미지 등록은 userService, userRoute에 정의하는 것이 좋을 듯 합니다.

Wonyocho replied:
userRoute 등록으로 수정 예정입니다.

Ares-10 commented on src/routes/user/uploadProfileImage.ts:
마찬가지로 라우터를 user.ts 로 옮기는 게 좋을 듯 합니다.
또한 이 파일은 API 경로와 데이터를 제공하는 라우터인데
- userRouter : API 경로 명시 및 인증된 유저 정보 받기, request 데이터 받기
- userService : 저장할 사진 이름, 파일, s3 디렉터리 위치 생성
- googleDrive : 저장할 사진 이름, 사진 파일, 저장할 s3디렉터리 위치만 받아서 실제로 구글 드라이브에 저장
으로 역할 및 책임 분리하는 게 좋을 듯 합니다.

Ares-10 commented on src/routes/user/uploadProfileImage.ts:
```diff
@@ -0,0 +1,69 @@
+        const data = Array.isArray(body.image) ? body.image[0] : body.image;
+
+        if (!data) {
+          throw fastify.httpErrors.badRequest('파일이 업로드되지 않았습니다.');
```
오류처리는 throw new GlobalException(GlobalErrorCode.에러코드이름)으로 하시는 게 좋을 것 같습니다. 전역 커스텀 오류처리 이용!

Ares-10 commented on src/plugins/user/userService.ts:
```diff
@@ -18,27 +18,25 @@ export default fp(async (fastify: FastifyInstance) => {
 
     /**
-     * 사용자 프로필 수정
+     * 사용자 이름 수정 (중복 확인 포함)
```
("중복 확인 포함") 주석은 사용자 이름 수정 함수에 암묵적으로 포함된 로직이라 주석으로 명시하지 않으셔도 될 듯 합니다.

Ares-10 commented on src/plugins/googleDrive.ts:
```diff
@@ -0,0 +1,44 @@
+      const auth = new google.auth.GoogleAuth({
+        credentials: {
+          client_email: GOOGLE_CLIENT_EMAIL,
+          private_key: GOOGLE_PRIVATE_KEY.replace(/\\n/g, '\n'),
```
.replace(/\\n/g, '\n') 부분은 무슨 용도인가요??

Wonyocho replied:
.replace(/\\n/g, '\n')는 문자열에서 이스케이프된 줄바꿈 문자(\n)를 실제 줄바꿈 문자로 변환하기 위함입니다.

Ares-10 commented on src/plugins/googleDrive.ts:
```diff
@@ -0,0 +1,44 @@
+
+export default fp(
+  async (fastify: FastifyInstance) => {
+    if (!GOOGLE_CLIENT_EMAIL) {
```
이 부분은 환경변수가 없을 때 서버 인스턴스 시작 부분에서 오류가 발생하니 불필요한 예외처리인 것 같습니다

Wonyocho replied:
환경변수 검증은 그러면 main.ts에서 하려고 하는데 어떻게 생각하시나요?

Commit:
refac: 사용자 이미지 업로드 기능 리팩토링 및 역할 분리

- userRoute에서 이미지 업로드 로직을 userService와 googleDriveService로 분리
- userService에 uploadUserImage 메서드 추가
- googleDriveService 구현: Google Drive 파일 업로드 및 공개 URL 생성
- Fastify 타입 정의 확장: userService 및 googleDriveService 메서드 추가
- main.ts에서 불필요한 라우트 제거 및 CORS 설정 위치 조정
- GlobalException에 새로운 에러 코드 추가

Commit:
fix: package-lock.json 의존성 업데이트 및 불필요한 항목 제거
